File: packages/sdk/src/interop.ts
export type InteropStatusEvent = {
  status: string;
  detail: string;
  created_at: string;
};

export type InteropMessage = {
  id: number;
  external_system: string;
  status: string;
  attempts: number;
  external_id: string;
  simulator_mode: boolean;
  created_at: string;
  status_events: InteropStatusEvent[];
};

export type InteropListResponse = {
  results: InteropMessage[];
};

export type InteropClientOptions = {
  baseUrl: string;
  fetch?: typeof fetch;
};

export function createInteropClient(options: InteropClientOptions) {
  const fetcher = options.fetch ?? fetch;
  const baseUrl = options.baseUrl.replace(/\/$/, "");

  async function request<T>(path: string, init?: RequestInit): Promise<T> {
    const response = await fetcher(`${baseUrl}${path}`, init);
    if (!response.ok) {
      const payload = await response.json().catch(() => ({}));
      const requestId = response.headers.get("x-request-id");
      const detail = payload.detail ?? `Request failed: ${response.status}`;
      throw new Error(requestId ? `${detail} (request_id: ${requestId})` : detail);
    }
    return response.json() as Promise<T>;
  }

  return {
    listMessages: () => request<InteropListResponse>("/interop/messages/"),
    processOutbox: () =>
      request<{ processed: number }>("/interop/process/", { method: "POST" }),
  };
}

File: apps/api/core/views/interop.py
from __future__ import annotations

import json

from django.conf import settings
from django.http import JsonResponse

from ..interop import SimulatorAdapter, enqueue_message, process_outbox
from ..models import InteropMessage
from ..rbac import has_permission


def interop_messages(request):
    membership = request.membership  # type: ignore[attr-defined]
    permission = has_permission(membership.role, "org:manage")
    if not permission.allowed:
        return JsonResponse({"detail": "Not authorized."}, status=403)
    if request.method == "POST":
        payload = json.loads(request.body or "{}")
        external_system = str(payload.get("external_system", "")).strip()
        if not external_system:
            return JsonResponse({"detail": "external_system required"}, status=400)
        simulator_mode = bool(payload.get("simulator", False))
        if simulator_mode and not settings.INTEROP_SIMULATOR_ENABLED:
            return JsonResponse({"detail": "Simulator not enabled"}, status=400)
        message = InteropMessage.objects.create(
            organization=membership.organization,
            episode_id=payload.get("episode_id"),
            external_system=external_system,
            payload=payload.get("payload", {}),
            status="draft",
            simulator_mode=simulator_mode,
            created_by=request.user,
        )
        enqueue_message(message)
        return JsonResponse({"id": message.id, "status": message.status}, status=201)
    messages = (
        InteropMessage.objects.filter(organization=membership.organization)
        .order_by("-created_at")[:100]
    )
    payload = [
        {
            "id": message.id,
            "external_system": message.external_system,
            "status": message.status,
            "attempts": message.attempts,
            "external_id": message.external_id,
            "simulator_mode": message.simulator_mode,
            "created_at": message.created_at.isoformat(),
            "status_events": [
                {
                    "status": event.status,
                    "detail": event.detail,
                    "created_at": event.created_at.isoformat(),
                }
                for event in message.status_events.order_by("created_at")
            ],
        }
        for message in messages
    ]
    return JsonResponse({"results": payload})


def interop_process(request):
    membership = request.membership  # type: ignore[attr-defined]
    permission = has_permission(membership.role, "org:manage")
    if not permission.allowed:
        return JsonResponse({"detail": "Not authorized."}, status=403)
    queued = InteropMessage.objects.filter(
        organization=membership.organization, status="queued"
    )
    if not settings.INTEROP_SIMULATOR_ENABLED:
        return JsonResponse({"detail": "Simulator not enabled"}, status=400)
    adapter = SimulatorAdapter(system_name="SIM")
    processed = process_outbox(list(queued), adapter)
    return JsonResponse({"processed": processed})
